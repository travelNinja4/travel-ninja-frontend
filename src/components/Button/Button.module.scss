/**
------------------------------------------------------
styles specific to the Button component
------------------------------------------------------
**/

.container {
  padding: 1rem;
  // Add styles here
}
