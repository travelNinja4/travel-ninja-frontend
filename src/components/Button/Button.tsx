/**
 * A reusable and customizable button component that supports various styles, sizes, and interaction states.
 *
 * @example
 * ```tsx
 * import Button from '@src/components/Button'
 *
 * export default function Button() {
 *   <Button startIcon={Plus} endIcon={ArrowRight} iconSize={18} iconColor="#333">
 *      Hello
 *    </Button>
 * }
 * ```
 */

import Typography from '../Typography';
import { LucideIcon } from 'lucide-react';
import clsx from 'clsx';
import styles from './Button.module.scss';

/**
 * Define the props available for the Button component.
 */
interface ButtonProps {
  /**
   * The content inside the button.
   * Typically a string or JSX (e.g., <span>Save</span>).
   */
  children: React.ReactNode;

  /**
   * Defines the behavior of the button when clicked.
   * - `button`: Standard clickable button (default).
   * - `submit`: Submits the form it belongs to.
   * - `reset`: Resets all form fields.
   *
   * @default "button"
   */
  type?: 'button' | 'submit' | 'reset';

  /**
   * If `true`, disables the button and prevents user interaction.
   *
   * @default false
   */
  disabled?: boolean;

  /**
   * Callback function triggered when the button is clicked.
   *
   * @param e The mouse click event.
   */
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;

  /**
   * Icon component (from `lucide-react`) rendered at the **start** of the button.
   */
  startIcon?: LucideIcon;

  /**
   * Icon component (from `lucide-react`) rendered at the **end** of the button.
   */
  endIcon?: LucideIcon;

  /**
   * Size of the start/end icons (in pixels).
   * Passed directly to the Lucide icon component.
   */
  iconSize?: number;

  /**
   * Color of the icons (CSS color value).
   * Accepts named colors, hex, rgb, or CSS variables.
   */
  iconColor?: string;

  /**
   * Additional CSS class names for customizing the button styles.
   * Useful for applying variant-specific styles.
   */
  className?: string;

  /**
   * If `true`, shows a loading spinner inside the button
   * and disables interaction.
   */
  loading?: boolean;
}

export default function Button({
  children,
  type = 'button',
  disabled = false,
  onClick,
  startIcon: StartIcon,
  endIcon: EndIcon,
  iconSize,
  iconColor,
  className,
  loading = false,
  ...rest
}: ButtonProps) {
  return (
    <button
      data-testid="ButtonTest"
      type={type}
      disabled={disabled}
      onClick={onClick}
      className={clsx(styles.button, className)}
      {...rest}
    >
      {loading ? (
        <>
          <span className={styles.loader} />
          <Typography tag="span" className={className}>
            {children}
          </Typography>
        </>
      ) : (
        <>
          {StartIcon && (
            <span className={styles.icon}>
              <StartIcon size={iconSize} color={iconColor} />
            </span>
          )}
          <Typography tag="span" className={className}>
            {children}
          </Typography>
          {EndIcon && (
            <span className={styles.icon}>
              <EndIcon size={iconSize} color={iconColor} />
            </span>
          )}
        </>
      )}
    </button>
  );
}
