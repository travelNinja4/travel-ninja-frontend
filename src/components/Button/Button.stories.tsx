import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { fn } from 'storybook/test';
import Button from './Button';

const meta = {
  title: 'components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    type: {
      control: { type: 'select' },
      options: ['button', 'submit', 'reset'],
    },
    disabled: {
      control: 'boolean',
    },
    children: {
      control: 'text',
    },
  },
  args: {
    children: 'Click Me',
    onClick: fn(),
    disabled: false,
    type: 'button',
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const WithStartIcon: Story = {
  args: {
    children: 'Save',
  },
};

export const Disabled: Story = {
  args: {
    children: 'Disabled',
    disabled: true,
  },
};
