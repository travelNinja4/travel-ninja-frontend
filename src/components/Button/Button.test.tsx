import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import Button from './Button'

describe('Home', () => {
  /** Base props for Button **/
  const baseProps={}
 
  /**
  * @test
  * Renders the Button component without throwing an error
  * and checks if the ButtonTest text is present in the document.
  *
  * This ensures that the component loads correctly with the base props.
  */
  it('renders without crashing', () => {
   render(<Button {...baseProps}/>);
   expect(screen.getByTestId('ButtonTest')).toBeInTheDocument();
  });

  /**
  * @test
  * Ensures that the Button component renders without throwing any runtime errors.
  *
  * This is a safety check to confirm that the component can mount successfully
  * with the provided base props, even if it does not render any visible content.
  */
  it('renders without throwing', () => {
  expect(() => render(<Button {...baseProps}/>)).not.toThrow();
  });

  /**
  * @test
  * Captures a snapshot of the Button component with base props
  * and ensures it matches the previously stored snapshot.
  *
  * This helps detect any unexpected UI changes.
  */
  it('matches snapshot', () => {
   const { asFragment } = render(<Button {...baseProps}/>);
   expect(asFragment()).toMatchSnapshot();
  });
})