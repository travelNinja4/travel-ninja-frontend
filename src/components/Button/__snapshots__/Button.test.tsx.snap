// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="button test-class"
    data-testid="ButtonTest"
    type="button"
  >
    <span
      class="typography span align-center test-class"
      data-testid="TypographyTest"
    >
      ButtonTest
    </span>
  </button>
</DocumentFragment>
`;
